File: ABC426/abc426_c.py
# ABC426 C - Upgrade Required
# See problem: https://atcoder.jp/contests/abc426/tasks/abc426_c
# 2025-10-04 TLE->AC+

import sys
from array import array

data = list(map(int, sys.stdin.buffer.read().split()))
it = iter(data)
N = next(it)
Q = next(it)

a = array('i', [1]) * (N + 1)
a[0] = 0

parent = array('i', range(N + 1))

def find(x: int) -> int:
    while x != parent[x]:
        parent[x] = parent[parent[x]]
        x = parent[x]
    return x

out_lines = []
for _ in range(Q):
    x = next(it); y = next(it)
    s = 0
    while True:
        p = find(x)
        if p == 0:
            break
        s += a[p]
        a[p] = 0
        parent[p] = find(p - 1)
    a[y] += s
    out_lines.append(str(s))

sys.stdout.write("\n".join(out_lines) + "\n")

File: ABC426/abc426_c.R
# ABC426 C - Upgrade Required
# See problem: https://atcoder.jp/contests/abc426/tasks/abc426_c
# 2025-10-04 TLE->AC-

options(scipen = 100, digits = 15)
input <- scan("stdin", what = integer(), quiet = TRUE)
N <- input[1]
Q <- input[2]

a <- rep.int(1, N + 1)
a[1] <- 0
parent <- 0:N + 1

out <- integer(Q)
idx <- 3

for (q in 1:Q) {
  x <- input[idx]
  idx <- idx + 1
  y <- input[idx]
  idx <- idx + 1

  s <- 0
  repeat {
    xi <- x + 1
    while (xi != parent[xi]) {
      parent[xi] <- parent[parent[xi]]
      xi <- parent[xi]
    }
    p <- xi - 1

    if (p == 0) break
    pi <- p + 1
    s <- s + a[pi]
    a[pi] <- 0
    yi <- p
    while (yi != parent[yi]) {
      parent[yi] <- parent[parent[yi]]
      yi <- parent[yi]
    }
    parent[pi] <- yi
  }
  a[y + 1] <- a[y + 1] + s
  out[q] <- s
}

cat(paste(out, collapse = "\n"), "\n")
